registry: change default registry config from `registry.json` to `registry/config.json`

zonggen opened the pull request:
Changes default registry config file from `registry.json` to `config.json`. This aligns with the `config.json` that is widely used in docker.

Closes: https://github.com/helm/helm/issues/10156
Related: https://github.com/helm/helm/issues/10122
Signed-off-by: Allen Bai <abai@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

Commit:
HELM_REGISTRY_CONFIG: change default file of registry config to config.json

Changes default registry config file from `registry.json` to
`config.json`. This aligns with the `config.json` that is widely used in docker.

Closes: https://github.com/helm/helm/issues/10156
Related: https://github.com/helm/helm/issues/10122
Signed-off-by: Allen Bai <abai@redhat.com>

mattfarina reviewed (commented):
Following what @jdolitsky said on #10122 :

> On working on another project involving Docker credentials, the filename config.json is hardcoded everywhere in Docker codebase.. It was a mistake to rename this file to registry.json. Should probably instead do something like /home/myname/.config/helm/.docker/config.json, which would allow people to use DOCKER_CONFIG=/home/myname/.config/helm/.docker

This all sorta makes sense. But, having the file in `HELM_CONFIG_PATH/config.json` is a bit confusing because it doesn't signal it's just for the registry. Would it be possible to move it to `HELM_CONFIG_PATH/registry/config.json`. This mirrors the use of `.docker` in the example from @jdolitsky

jdolitsky commented:
THE PR in its current state works to solve the issue, but I think one of the following would be more clear:

- `HELM_CONFIG_PATH/registry/config.json` (as @mattfarina suggests)
- `HELM_CONFIG_PATH/.docker/config.json`

Ideally (but can also be later) `helm env` could return the path to this dir in `HELM_REGISTRY_DOCKER_CONFIG` (or similar) so we could use it, for example:

```
$ export DOCKER_CONFIG="$(helm env | grep HELM_REGISTRY_DOCKER_CONFIG | cut -d "=" -f2 | tr -d '"')"
$ echo "DOCKER_CONFIG: ${DOCKER_CONFIG}"
DOCKER_CONFIG: /Users/myuser/Library/Preferences/helm/registry
```

jdolitsky commented:
The `HELM_REGISTRY_CONFIG` env is mostly useless, and perhaps we can just re-use that to point to the directory (vs. full path to config.json)

scottrigby commented:
I know we don't *need* to, but keeping in mind existing users of the experimental feature – would the principle of least astonishment have us choose a new var/flag (like `--registry-config-dir` or `--registry_config_home` and corresponding env vars, and throw a deprecation warning for `HELM_REGISTRY_CONFIG`? 

Or would it be equally clear/confusing/easy/annoying to existing users to re-use the same flag and var but now pointing to the dir rather than file? I have not really used this feature so it's not easy for me to put myself in that mindset. 

I am thinking though, either way, I'd want my existing setup from of experimental feature to "just work" when the feature is made into a full one. WDYT?

scottrigby commented:
@zonggen do you want to connect on this in the morning?

zonggen commented:
@scottrigby Definitely, will ping you on Slack.

Thank you for the feedbacks. I echo the idea of `HELM_CONFIG_PATH/registry/config.json` and/or `HELM_CONFIG_PATH/.docker/config.json`. Though I had the same questions as Scott during implementation and wondered what the scope of this PR should be.

scottrigby commented:
@zonggen see current implementation here https://github.com/helm/helm/blob/main/cmd/helm/root.go#L155

jdolitsky commented:
@scottrigby sure. I dont think there is a problem with keeping `HELM_REGISTRY_CONFIG` as is. Nice shell foo.

If we want to expose the directory later, we can use _DIR or add DOCKER to a new var

Commit:
registry/client: use NewClientWithDockerFallback from oras v1.1.0-rc3

Uses NewClientWithDockerFallback to always use Docker config and
modifies go.mod to use oras v1.1.0-rc3.

Ref:
 - https://github.com/helm/helm/pull/10536#issuecomment-1010144075
 - https://github.com/oras-project/oras-go/pull/92
Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented:
Thank you @scottrigby for the help! I've updated this PR with the followings:
- [x] Change `registry.json` to `config.json` for `CredentialsFileBasename` and `registry.json` to `registry/config.json` for `RegistryConfig`.
- [x] Use `NewClientWithDockerFallback` instead of `NewClient` in `registry/client.go`.
- [x] Update to `oras.land/oras-go v1.1.0-rc3` in `go.mod`.

Please take a look and let me know if this looks sane.

jdolitsky commented:
@zonggen - I think im missing the issue here. `helm registry login` will always ask for password. This is how you might overwrite credentials.

For testing purposes, I recommend *not* using Docker Hub, as they do not support Helm charts. Use Azure/Amazon/Google or GitHub, or probably better just stand a `localhost:5000` registry with auth. Here are some tips for how to do the last suggestion: https://helm.sh/docs/topics/registries/#auth

jdolitsky commented:
"does it fall back on Docker config properly?"
Yes. Here's what I did to test:
```
$ docker login bundle.bar
Username: myuser
Password: 
Login Succeeded

$ helm create mychart
Creating mychart

$ helm package mychart/
Successfully packaged chart and saved it to: /Users/myuser/mychart-0.1.0.tgz

$ export HELM_EXPERIMENTAL_OCI=1

$ helm push ./mychart-0.1.0.tgz oci://bundle.bar/u/myuser/issue-10156
Pushed: bundle.bar/u/myuser/issue-10156/mychart:0.1.0
Digest: sha256:e5dd8ffd211be03e3eeb7b025a855b8d01a8f1104bf2a9a08a250fc4bd87e72a
```

jdolitsky commented:
Ok this PR is working as expected. Thank you very much @zonggen 

Helm now supports the `DOCKER_CONFIG` env var, as well as Docker's default location. Helm can rely on Docker's config, and Docker can rely on Helm's.

Example - push to GCR via [docker-credential-magic](https://github.com/docker-credential-magic/docker-credential-magic) cred helper:

```
# Setup
$ mkdir /tmp/helm-demo
$ echo '{"credsStore": "magic"}' > /tmp/helm-demo/config.json
$ export DOCKER_CONFIG=/tmp/helm-demo

# Helm v3.7.2
$ helm push ./mychart-0.1.0.tgz oci://gcr.io/my-gcp-project/helm-testing
Error: failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized

# Helm built from this changeset
$ bin/helm push ./mychart-0.1.0.tgz oci://gcr.io/my-gcp-project/helm-testing
Pushed: gcr.io/my-gcp-project/helm-testing/mychart:0.1.0
Digest: sha256:e5dd8ffd211be03e3eeb7b025a855b8d01a8f1104bf2a9a08a250fc4bd87e72a
```

Example - Docker pointing to Helm's config:

```
# Setup (overwrites all helm logins with magic helper)
$ export DOCKER_CONFIG=`dirname $(helm env | grep HELM_REGISTRY_CONFIG | cut -d "=" -f2 | tr -d '"')`
$ echo '{"credsStore": "magic"}' > "${DOCKER_CONFIG}/config.json"

# Download sample image
$ docker pull alpine
Using default tag: latest
latest: Pulling from library/alpine
9b3977197b4f: Pull complete 
Digest: sha256:21a3deaa0d32a8057914f36584b5288d2e5ecc984380bc0118285c70fa8c9300
Status: Downloaded newer image for alpine:latest
docker.io/library/alpine:latest

# Tag image for private registry
$ docker tag alpine gcr.io/my-gcp-project/helm-testing-docker:1

# Push succeeds
$ docker push gcr.io/my-gcp-project/helm-testing-docker:1
The push refers to repository [gcr.io/my-gcp-project/helm-testing-docker]
07d3c46c9599: Layer already exists 
1: digest: sha256:c74f1b1166784193ea6c8f9440263b9be6cae07dfe35e32a5df7a31358ac2060 size: 528

# Push fails after unsetting DOCKER_CONFIG
$ unset DOCKER_CONFIG
$ docker push gcr.io/my-gcp-project/helm-testing-docker:1
The push refers to repository [gcr.io/my-gcp-project/helm-testing-docker]
07d3c46c9599: Preparing 
unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication
```

Example - basic auth registry:
```
# Remove the Helm config.json
rm -f $(bin/helm env | grep HELM_REGISTRY_CONFIG | cut -d "=" -f2 | tr -d '"')

# Login
$ bin/helm registry login bundle.bar
Username: myuser
Password: 
Login Succeeded

# Show config file
$ cat $(bin/helm env | grep HELM_REGISTRY_CONFIG | cut -d "=" -f2 | tr -d '"')
{
	"auths": {
		"bundle.bar": {}
	},
	"credsStore": "osxkeychain"
}

# Push (succeeds)
$ bin/helm push ./mychart-0.1.0.tgz oci://bundle.bar/u/myuser/helm-demo-1
Pushed: bundle.bar/u/myuser/helm-demo-1/mychart:0.1.0
Digest: sha256:e5dd8ffd211be03e3eeb7b025a855b8d01a8f1104bf2a9a08a250fc4bd87e72a

# Logout
$ bin/helm registry logout bundle.bar
Removing login credentials for bundle.bar

# Show config file again (now missing entry)
$ cat $(bin/helm env | grep HELM_REGISTRY_CONFIG | cut -d "=" -f2 | tr -d '"')
{
	"auths": {},
	"credsStore": "osxkeychain"
}

# Push (fails)
$ bin/helm push ./mychart-0.1.0.tgz oci://bundle.bar/u/myuser/helm-demo-2
Error: failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized
```

Good stuff

zonggen commented:
Thank you for the thorough example walkthrough @jdolitsky :))
